exchanges/exchanges.go

```diff
@@ -71,15 +160,195 @@ func Tokens() []string {
+		Price:      state.Price,
+		BaseVolume: state.BaseVolume,
+		Volume:     state.Volume,
+		Change:     state.Change,
+		Stamp:      state.Stamp,
+		Depth:      state.Depth,
+	}
+	if state.Candlesticks != nil {
+		newState.Candlesticks = make(map[candlestickKey]Candlesticks)
+		for bin, sticks := range state.Candlesticks {
+			newState.Candlesticks[bin] = sticks
+		}
+	}
+	return newState
+}
+
+// Create a new exchange state that projects the fields from `top` over a copy
+// of the receiver.
+func (state *ExchangeState) project(top *ExchangeState) *ExchangeState {
+	// Starting with a copy of top since the value fields must be copied from
+	// the top anyway.
+	bottom := top.copy()
+
+	// Start with the current set of candlesticks, and overwrite with any new
+	// entries.
+	var sticks map[candlestickKey]Candlesticks
+	if state != nil {
+		sticks = state.Candlesticks
+	}
+	if sticks == nil && top.Candlesticks != nil {
+		sticks = make(map[candlestickKey]Candlesticks)
+	}
+	for bin, newSticks := range top.Candlesticks {
+		sticks[bin] = newSticks
+	}
+	bottom.Candlesticks = sticks
+
+	// Careful not to overwrite with nil unless data is expired.
+	if bottom.Depth == nil && state.Depth != nil && state.Depth.IsFresh() {
+		bottom.Depth = state.Depth
+	}
+
+	return bottom
+}
+
+// Parse an ExchangeState from a protocol buffer message.
+func exchangeStateFromProto(proto *dcrrates.ExchangeRateUpdate) *ExchangeState {
+	state := &ExchangeState{
+		Price:      proto.GetPrice(),
+		BaseVolume: proto.GetBaseVolume(),
+		Volume:     proto.GetVolume(),
+		Change:     proto.GetChange(),
+		Stamp:      proto.GetStamp(),
+	}
+
+	updateDepth := proto.GetDepth()
+	if updateDepth != nil {
+		depth := &DepthData{
+			Time: time.Unix(updateDepth.Time, 0),
+			Bids: make([]DepthPoint, 0, len(updateDepth.Bids)),
+			Asks: make([]DepthPoint, 0, len(updateDepth.Asks)),
+		}
+		for _, bid := range updateDepth.Bids {
+			depth.Bids = append(depth.Bids, DepthPoint{
+				Quantity: bid.Quantity,
+				Price:    bid.Price,
+			})
+		}
+		for _, ask := range updateDepth.Asks {
+			depth.Asks = append(depth.Asks, DepthPoint{
+				Quantity: ask.Quantity,
+				Price:    ask.Price,
+			})
+		}
+		state.Depth = depth
+	}
+
+	if proto.Candlesticks != nil {
+		stickMap := make(map[candlestickKey]Candlesticks)
+		for _, candlesticks := range proto.Candlesticks {
+			sticks := make(Candlesticks, 0, len(candlesticks.Sticks))
+			for _, stick := range candlesticks.Sticks {
+				sticks = append(sticks, Candlestick{
+					High:   stick.High,
+					Low:    stick.Low,
+					Open:   stick.Open,
+					Close:  stick.Close,
+					Volume: stick.Volume,
+					Start:  time.Unix(stick.Start, 0),
+				})
+			}
+			stickMap[candlestickKey(candlesticks.Bin)] = sticks
+		}
+		state.Candlesticks = stickMap
+	}
+	return state
+}
+
+// HasCandlesticks checks for data in the candlesticks map.
+func (state *ExchangeState) HasCandlesticks() bool {
+	return len(state.Candlesticks) > 0
```

Comment:
I notice you check in other functions for `state != nil`.  Is there any expected way that will happen (for those functions or these)?

Reply:
That was a little weird, but I think I resolved it when I removed `project`.